File: static/js/game_js_modules/constants_js_modules/vimMotionScoring.js
// Vim Motion Scoring System
// This mirrors the backend scoring system for consistent UI feedback

const VIM_MOTION_SCORES = {
  // Basic directional movement (h, j, k, l)
  basicMovement: {
    "h": 100,
    "j": 100,
    "k": 100,
    "l": 100,
  },

  // Word movement (w, W, e, E, b, B, ge, gE)
  wordMovement: {
    "w": 120,
    "W": 120,
    "e": 120,
    "E": 120,
    "b": 120,
    "B": 120,
    "ge": 120,
    "gE": 120,
  },

  // Line movement (0, $, ^, g_)
  lineMovement: {
    "0": 120,
    "$": 120,
    "^": 130,
    "g_": 130,
  },

  // Find character movement (f, F, t, T)
  findChar: {
    "f": 125,
    "F": 125,
    "t": 125,
    "T": 125,
  },

  // Repeat find movement (; and ,)
  repeatFind: {
    ";": 130,
    ",": 130,
  },

  // File/line jumping (gg, G, line numbers)
  fileJump: {
    "gg": 130,
    "G": 130,
  },

  // Screen movement (H, M, L)
  screenMovement: {
    "H": 150,
    "M": 150,
    "L": 150,
  },

  // Paragraph/section movement ({, }, (, ))
  paragraphMovement: {
    "{": 160,
    "}": 160,
    "(": 160,
    ")": 160,
  },

  // Match movement (%)
  matchMovement: {
    "%": 200,
  },

  // Arrow key penalty
  arrowPenalty: {
    "ArrowUp": -50,
    "ArrowDown": -50,
    "ArrowLeft": -50,
    "ArrowRight": -50,
  },
};

// Get score for a given motion
export function getMotionScore(motion) {
  // Check each category for the motion
  if (VIM_MOTION_SCORES.basicMovement[motion]) {
    return VIM_MOTION_SCORES.basicMovement[motion];
  }
  if (VIM_MOTION_SCORES.wordMovement[motion]) {
    return VIM_MOTION_SCORES.wordMovement[motion];
  }
  if (VIM_MOTION_SCORES.lineMovement[motion]) {
    return VIM_MOTION_SCORES.lineMovement[motion];
  }
  if (VIM_MOTION_SCORES.findChar[motion]) {
    return VIM_MOTION_SCORES.findChar[motion];
  }
  if (VIM_MOTION_SCORES.repeatFind[motion]) {
    return VIM_MOTION_SCORES.repeatFind[motion];
  }
  if (VIM_MOTION_SCORES.fileJump[motion]) {
    return VIM_MOTION_SCORES.fileJump[motion];
  }
  if (VIM_MOTION_SCORES.screenMovement[motion]) {
    return VIM_MOTION_SCORES.screenMovement[motion];
  }
  if (VIM_MOTION_SCORES.paragraphMovement[motion]) {
    return VIM_MOTION_SCORES.paragraphMovement[motion];
  }
  if (VIM_MOTION_SCORES.matchMovement[motion]) {
    return VIM_MOTION_SCORES.matchMovement[motion];
  }
  if (VIM_MOTION_SCORES.arrowPenalty[motion]) {
    return VIM_MOTION_SCORES.arrowPenalty[motion];
  }

  // Handle numbered G commands (e.g., "5G")
  if (motion.length > 1 && motion[motion.length - 1] === 'G') {
    // Check if it's a numbered G command
    const numberPart = motion.slice(0, -1);
    if (/^\d+$/.test(numberPart)) {
      return VIM_MOTION_SCORES.fileJump["G"];
    }
  }

  // Handle find character motions (e.g., "find_char_forward_a")
  if (motion.startsWith("find_char_forward_")) {
    return VIM_MOTION_SCORES.findChar["f"];
  }
  if (motion.startsWith("find_char_backward_")) {
    return VIM_MOTION_SCORES.findChar["F"];
  }
  if (motion.startsWith("till_char_forward_")) {
    return VIM_MOTION_SCORES.findChar["t"];
  }
  if (motion.startsWith("till_char_backward_")) {
    return VIM_MOTION_SCORES.findChar["T"];
  }

  // Default to basic movement score for unknown motions
  return VIM_MOTION_SCORES.basicMovement["h"];
}

// Get motion category for a given motion
export function getMotionCategory(motion) {
  if (VIM_MOTION_SCORES.basicMovement[motion]) {
    return "Basic Movement";
  }
  if (VIM_MOTION_SCORES.wordMovement[motion]) {
    return "Word Movement";
  }
  if (VIM_MOTION_SCORES.lineMovement[motion]) {
    return "Line Movement";
  }
  if (VIM_MOTION_SCORES.findChar[motion]) {
    return "Find Character";
  }
  if (VIM_MOTION_SCORES.repeatFind[motion]) {
    return "Repeat Find";
  }
  if (VIM_MOTION_SCORES.fileJump[motion]) {
    return "File Jump";
  }
  if (VIM_MOTION_SCORES.screenMovement[motion]) {
    return "Screen Movement";
  }
  if (VIM_MOTION_SCORES.paragraphMovement[motion]) {
    return "Paragraph Movement";
  }
  if (VIM_MOTION_SCORES.matchMovement[motion]) {
    return "Match Movement";
  }
  if (VIM_MOTION_SCORES.arrowPenalty[motion]) {
    return "Arrow Penalty";
  }

  // Handle numbered G commands
  if (motion.length > 1 && motion[motion.length - 1] === 'G') {
    const numberPart = motion.slice(0, -1);
    if (/^\d+$/.test(numberPart)) {
      return "File Jump";
    }
  }

  // Handle find character motions
  if (motion.startsWith("find_char_forward_") || motion.startsWith("till_char_forward_") ||
      motion.startsWith("find_char_backward_") || motion.startsWith("till_char_backward_")) {
    return "Find Character";
  }

  return "Basic Movement";
}

// Get formatted score display text
export function getScoreDisplay(motion) {
  const score = getMotionScore(motion);
  const category = getMotionCategory(motion);
  
  return {
    score: score,
    category: category,
    text: score > 0 ? `+${score}` : `${score}`,
  };
}

File: internal/constant/vim_motion_scoring.go
package constant

// VimMotionScores defines the scoring system for different vim motions
type VimMotionScores struct {
	// Basic directional movement (h, j, k, l)
	BasicMovement map[string]int

	// Word movement (w, W, e, E, b, B, ge, gE)
	WordMovement map[string]int

	// Line movement (0, $, ^, g_)
	LineMovement map[string]int

	// Find character movement (f, F, t, T)
	FindChar map[string]int

	// Repeat find movement (; and ,)
	RepeatFind map[string]int

	// File/line jumping (gg, G, line numbers)
	FileJump map[string]int

	// Screen movement (H, M, L)
	ScreenMovement map[string]int

	// Paragraph/section movement ({, }, (, ))
	ParagraphMovement map[string]int

	// Match movement (%)
	MatchMovement map[string]int

	// Arrow key penalty
	ArrowPenalty map[string]int
}

// GetVimMotionScores returns the configured scoring system
func GetVimMotionScores() *VimMotionScores {
	return &VimMotionScores{
		BasicMovement: map[string]int{
			"h": 100,
			"j": 100,
			"k": 100,
			"l": 100,
		},
		WordMovement: map[string]int{
			"w":  120,
			"W":  120,
			"e":  120,
			"E":  120,
			"b":  120,
			"B":  120,
			"ge": 120,
			"gE": 120,
		},
		LineMovement: map[string]int{
			"0":  120,
			"$":  120,
			"^":  130,
			"g_": 130,
		},
		FindChar: map[string]int{
			"f": 125,
			"F": 125,
			"t": 125,
			"T": 125,
		},
		RepeatFind: map[string]int{
			";": 130,
			",": 130,
		},
		FileJump: map[string]int{
			"gg": 130,
			"G":  130,
		},
		ScreenMovement: map[string]int{
			"H": 150,
			"M": 150,
			"L": 150,
		},
		ParagraphMovement: map[string]int{
			"{": 160,
			"}": 160,
			"(": 160,
			")": 160,
		},
		MatchMovement: map[string]int{
			"%": 200,
		},
		ArrowPenalty: map[string]int{
			"ArrowUp":    -50,
			"ArrowDown":  -50,
			"ArrowLeft":  -50,
			"ArrowRight": -50,
		},
	}
}

// GetMotionScore returns the score for a given motion
func (vms *VimMotionScores) GetMotionScore(motion string) int {
	// Check each category for the motion
	if score, exists := vms.BasicMovement[motion]; exists {
		return score
	}
	if score, exists := vms.WordMovement[motion]; exists {
		return score
	}
	if score, exists := vms.LineMovement[motion]; exists {
		return score
	}
	if score, exists := vms.FindChar[motion]; exists {
		return score
	}
	if score, exists := vms.RepeatFind[motion]; exists {
		return score
	}
	if score, exists := vms.FileJump[motion]; exists {
		return score
	}
	if score, exists := vms.ScreenMovement[motion]; exists {
		return score
	}
	if score, exists := vms.ParagraphMovement[motion]; exists {
		return score
	}
	if score, exists := vms.MatchMovement[motion]; exists {
		return score
	}
	if score, exists := vms.ArrowPenalty[motion]; exists {
		return score
	}

	// Handle numbered G commands (e.g., "5G")
	if len(motion) > 1 && motion[len(motion)-1] == 'G' {
		// Check if it's a numbered G command
		for i := 0; i < len(motion)-1; i++ {
			if motion[i] < '0' || motion[i] > '9' {
				break
			}
			if i == len(motion)-2 {
				// It's a valid numbered G command
				return vms.FileJump["G"]
			}
		}
	}

	// Handle find character motions (e.g., "find_char_forward_a")
	if len(motion) > 18 && motion[:18] == "find_char_forward_" {
		return vms.FindChar["f"]
	}
	if len(motion) > 19 && motion[:19] == "find_char_backward_" {
		return vms.FindChar["F"]
	}
	if len(motion) > 18 && motion[:18] == "till_char_forward_" {
		return vms.FindChar["t"]
	}
	if len(motion) > 19 && motion[:19] == "till_char_backward_" {
		return vms.FindChar["T"]
	}

	// Default to basic movement score for unknown motions
	return vms.BasicMovement["h"]
}

// GetMotionCategory returns the category name for a given motion
func (vms *VimMotionScores) GetMotionCategory(motion string) string {
	if _, exists := vms.BasicMovement[motion]; exists {
		return "Basic Movement"
	}
	if _, exists := vms.WordMovement[motion]; exists {
		return "Word Movement"
	}
	if _, exists := vms.LineMovement[motion]; exists {
		return "Line Movement"
	}
	if _, exists := vms.FindChar[motion]; exists {
		return "Find Character"
	}
	if _, exists := vms.RepeatFind[motion]; exists {
		return "Repeat Find"
	}
	if _, exists := vms.FileJump[motion]; exists {
		return "File Jump"
	}
	if _, exists := vms.ScreenMovement[motion]; exists {
		return "Screen Movement"
	}
	if _, exists := vms.ParagraphMovement[motion]; exists {
		return "Paragraph Movement"
	}
	if _, exists := vms.MatchMovement[motion]; exists {
		return "Match Movement"
	}
	if _, exists := vms.ArrowPenalty[motion]; exists {
		return "Arrow Penalty"
	}

	// Handle numbered G commands
	if len(motion) > 1 && motion[len(motion)-1] == 'G' {
		for i := 0; i < len(motion)-1; i++ {
			if motion[i] < '0' || motion[i] > '9' {
				break
			}
			if i == len(motion)-2 {
				return "File Jump"
			}
		}
	}

	// Handle find character motions
	if len(motion) > 18 && (motion[:18] == "find_char_forward_" || motion[:18] == "till_char_forward_") {
		return "Find Character"
	}
	if len(motion) > 19 && (motion[:19] == "find_char_backward_" || motion[:19] == "till_char_backward_") {
		return "Find Character"
	}

	return "Basic Movement"
}